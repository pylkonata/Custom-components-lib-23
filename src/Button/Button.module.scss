@import url('https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,300;0,400;0,500;0,700;0,900;1,300;1,400;1,500;1,700&display=swap');

.btn {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  outline: 0;
  border: 0;
  margin: 0;
  cursor: pointer;
  user-select: none;
  text-decoration: none;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  font-weight: 500;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  border-radius: 4px;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.small {
  font-size: 0.8125rem;
}

.medium {
  font-size: 0.875rem;
}

.large {
  font-size: 0.9375rem;
}

.text {
  background-color: transparent;
  color: #1976d2;

  &:hover {
    background-color: rgba(25, 118, 210, 0.04);
  }

  &.small {
    padding: 4px 5px;
  }

  &.medium {
    padding: 6px 8px;
  }

  &.large {
    padding: 8px 11px;
  }

  &:disabled {
    color: rgba(0, 0, 0, 0.26);
    pointer-events: none;
    cursor: default;
  }
}


.contained {
  color: #fff;
  background-color: #1976d2;
  box-shadow: 0px 3px 1px -2px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 1px 5px 0px rgba(0, 0, 0, 0.12);

  &.small {
    padding: 4px 10px;
  }

  &.medium {
    padding: 6px 16px;
  }

  &.large {
    padding: 8px 22px;
  }

  &:hover {
    background-color: #1565c0;
    box-shadow: 0px 2px 4px -1px rgba(0, 0, 0, 0.2), 0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12);
  }

  &:disabled {
    color: rgba(0, 0, 0, 0.26);
    box-shadow: none;
    background-color: rgba(0, 0, 0, 0.12);
    pointer-events: none;
    cursor: default;
  }
}


.outlined {
  border: 1px solid rgba(25, 118, 210, 0.5);
  color: #1976d2;
  background-color: transparent;

  &:hover {
    background-color: rgba(25, 118, 210, 0.04);
    border: 1px solid #1976d2;
  }

  &.small {
    padding: 3px 9px;
  }

  &.medium {
    padding: 5px 15px;
  }

  &.large {
    padding: 7px 21px;
  }

  &:disabled {
    color: rgba(0, 0, 0, 0.26);
    border: 1px solid rgba(0, 0, 0, 0.12);
    pointer-events: none;
    cursor: default;
  }
}
