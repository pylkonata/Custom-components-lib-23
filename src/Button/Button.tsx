import React from 'react';
import s from './Button.module.scss';

type VariantBtn = 'text' | 'contained' | 'outlined';
type SizeBtn = 'small' | 'medium' | 'large';

interface ButtonProps {
	variant: VariantBtn;
	size: SizeBtn;
	onClick?: () => void;
	disabled?: boolean;
	children: React.ReactNode;
}
const Button = ({
	variant = 'contained',
	size,
	onClick,
	disabled = false,
	children,
	...props
}: ButtonProps) => {
	const classStyles = [s.btn, s[variant], s[size]];
	return (
		<button className={classStyles.join(' ')} disabled={disabled} onClick={onClick} {...props}>
			{children}
		</button>
	);
};

export default Button;
