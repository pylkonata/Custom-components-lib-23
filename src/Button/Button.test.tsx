import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Button from './Button';

describe('Button', () => {
	test('render default button', () => {
		render(<Button size='medium' children='Default button' />);

		const btn = screen.getByRole('button');
		expect(btn).toBeInTheDocument();
		expect(btn).toMatchSnapshot();
		expect(btn).toBeEnabled();
		expect(btn).toHaveClass('btn', 'contained', 'medium');
	});

	test('render disabled button with not call onClick', () => {
		const handleClick = jest.fn();
		render(
			<Button
				variant='outlined'
				size='large'
				disabled={true}
				children='Disabled button'
				onClick={handleClick}
			/>
		);

		const btn = screen.getByRole('button');
		expect(btn).toBeInTheDocument();
		expect(btn).toMatchSnapshot();
		expect(btn).toBeDisabled();
		expect(btn).toHaveClass('btn', 'large', 'outlined');
		fireEvent.click(btn);
		expect(handleClick).not.toHaveBeenCalled();
	});

	test('render button with onClick function', () => {
		const handleClick = jest.fn();
		render(<Button variant='text' size='small' onClick={handleClick} children='Click button' />);

		const btn = screen.getByText(/click button/i);
		expect(btn).toBeInTheDocument();
		expect(btn).toHaveClass('btn', 'text', 'small');
		fireEvent.click(btn);
		expect(handleClick).toHaveBeenCalledTimes(1);
	});
});
