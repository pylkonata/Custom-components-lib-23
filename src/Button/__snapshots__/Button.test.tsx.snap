// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render default button 1`] = `
<button
  class="btn contained medium"
>
  Default button
</button>
`;

exports[`Button render disabled button with not call onClick 1`] = `
<button
  class="btn outlined large"
  disabled=""
>
  Disabled button
</button>
`;
